docs/fpc-key-dist.puml

```diff
@@ -1,97 +1,319 @@
+        This could enable key distribution from an old cc version,
+        with a different MRENCLAVE.
+        Notice: in this case there cannot be also a check
+        can_endorse(ECC1_S_PK) -- or it should be negated.
+    end note
+  return true /'Enclave1 -> TLCC_Enclave1'/
+    note left
+        if false, abort
+    end note
+  Enclave1      -> Enclave1         :   check CC_name, CC_ver, CH_id, org_id in sealed data
+  Enclave1      -> Enclave1         :   check ECC2_S_PK, ECC_E_PK, attestation_report
+  Enclave1      -> Enclave1         :   m <- encrypt <SEK, CC_E_SK> with ECC2_E_PK
+    note left
+        enclave encrypts state encryption key and chaincode private key
+        - m denotes encrypted message
+    end note
+
+  Enclave1      -> Enclave1         :   SIG_ECC1 <- sign <ECC1_S_PK, ECC2_s_PK, CC_E_PK, m> with ECC1_S_SK
+    note left
+        enclave signs the key material
+        it distributes together with the
+        sender and intended recipient identity
+    end note
+
+  Enclave1      -> Enclave1         :   w <- <ECC1_S_PK, ECC2_s_PK, CC_E_PK, SIG_ECC1>
+
+  return m, w /'ECC1 -> Enclave1'/
+  return m, w /'ECC_Manager1 -> ECC1'/
+  return m, w /'Admin1 -> ECC_Manager1'/
+
+  note right Admin1
+    Key distribution is in progress. We need a means to transfer the message with
+    the encrypted keys to the recipient enclave. The next step uses the ledger
+    itself for this data transfer. This additionally allows to log the two phases
+    of the key distribution: (1) the sender enclave produced the encrypted data blob;
+    (2) the recipient enclave correctly received it. This can be useful when dealing
+    with enclaves of two different orgs.
   end note
 
-  return ECC_S_PK, SIG_ECC, m, CC_PK
-  return ECC_S_PK, SIG_ECC, m, CC_PK
-  return ECC_S_PK, SIG_ECC, m, CC_PK
+  group start key distribution endorsement
+    Admin1      -> ERCC1            ++: Start_KeyDist(m, w)
+    ERCC1       -> ERCC1            :   check ECC1_S_PK, ECC2_s_PK are registered
+    ERCC1       -> TLCC_Enclave1++: can_endorse(ECC1_S_PK)
+    return true
+    ERCC1       -> TLCC_Enclave1++: can_endorse(ECC2_S_PK)
+    return true
+    ERCC1       -> ERCC1            :   check SIG_ECC1
+    ERCC1       -> ERCC1            :   put_state(key=<CC_name, CC_ver, ECC1_S_PK, "keydist_started">, value=<w, m>)
+    return /'Admin1 -> ERCC1'/
+  end group
+
+  ref over Admin1, ERCC2
+    start key distribution endorsement
+  end ref
+
+  Admin1 -> Orderer: submit start key distribution transaction
+end group
+
+group complete key distribution
+  Admin2        -> Admin2           :   m <- <CC_name, CC_ver, CH_id, org_id>
+  Admin2        -> ECC_Manager2     ++: ConfirmCCKeyDist(m)
+  ECC_Manager2  -> ECC2             ++: ConfirmCCKeyDist(m)
+  ECC2          -> Enclave2         ++: ConfirmCCKeyDist(m)
+  Enclave2      -> ERCC2            ++: GetKeyDistStart(key=<CC_name, CC_ver, ECC2_S_PK, "keydist_started">)
```

Comment:
I don't like the term `start` as it makes me looking for keydist_ended :)

ouh there is a complete.

Reply:
At this time, when ECC2 has to retrieve the shared_CCKeys which was produced by ECC1 and published on the ledger, it produces a commitment that it has completed the key distribution.

Reply:
it's not a commitment, it's a confirmation